src/main/java/org/traccar/WebDataHandler.java

```diff
@@ -155,32 +181,95 @@ public String formatRequest(Position position) throws UnsupportedEncodingExcepti
 
-        String url;
-        if (json) {
-            url = this.url;
-        } else {
-            try {
-                url = formatRequest(position);
-            } catch (UnsupportedEncodingException | JsonProcessingException e) {
-                throw new RuntimeException("Forwarding formatting error", e);
+        class AsyncRequestAndCallback implements InvocationCallback<Response> {
+
+            private int delay = retryDelayMin;
+            private Map<String, Object> payload;
+            private Invocation.Builder requestBuilder;
+
+            AsyncRequestAndCallback(Position position) {
+
+                String formattedUrl;
+                try {
+                    formattedUrl = json ? url : formatRequest(position);
+                } catch (UnsupportedEncodingException | JsonProcessingException e) {
+                    throw new RuntimeException("Forwarding formatting error", e);
+                }
+
+                requestBuilder = client.target(formattedUrl).request();
+                if (header != null && !header.isEmpty()) {
+                    for (String line: header.split("\\r?\\n")) {
+                        String[] values = line.split(":", 2);
+                        requestBuilder.header(values[0].trim(), values[1].trim());
+                    }
+                }
+
+                if (json) {
+                    payload = prepareJsonPayload(position);
+                }
+
+                deliveryPendingCurrent.incrementAndGet();
+
+                send();
             }
-        }
 
-        Invocation.Builder requestBuilder = client.target(url).request();
+            private void send() {
+                if (json) {
+                    requestBuilder.async().post(Entity.json(payload), this);
+                } else {
+                    requestBuilder.async().get(this);
+                }
+            }
 
-        if (header != null && !header.isEmpty()) {
-            for (String line: header.split("\\r?\\n")) {
-                String[] values = line.split(":", 2);
-                requestBuilder.header(values[0].trim(), values[1].trim());
+            private void retry() {
+                try {
+                    String message = "Position forwarding failed.";
+                    if (!retryEnabled) {
+                        LOGGER.warn(message);
+                    } else {
+                        int pending = deliveryPendingCurrent.get();
+                        if (pending <= deliveryPendingLimit) {
+                            LOGGER.warn(message + " Pending: " + pending
+                                + ". Retrying in " + delay + " seconds.");
+                            GlobalTimer.getTimer().newTimeout(new TimerTask() {
+                                @Override
+                                public void run(Timeout timeout) {
+                                    if (!timeout.isCancelled()) {
+                                        if (delay < retryDelayMax) {
```

Comment:
I thought the idea was that it will stop retying after it reaches max. Otherwise I don't see any reason to have a max. Also, usually interval is doubled, not incremented.

Reply:
The point is that I don't want positions to be delivered, lets say half an hour, after the problem that caused the failures, has been resolved (at minute 16), and even worst, newer positions to be dropped after the first try, just because the pending limit has been reached by positions that again, are waiting for half an hour to be delivered.
What if instead of a min and a max, we use a configurable retry interval, and only retry if a successful delivery happened while waiting? In this case the only limit will be the configurable value of how much pending deliveries we can have. If this limit is reached, the first delivery to be retried will be dropped (It's maybe not the oldest, but will never be the newest)
Maybe I'm seeing this wrong, but please, tell me what do you think.
Thanks.

Reply:
Standard way to implement something like this is retry X number of times with doubling interval. For example:

5s -> 10s -> 20s -> 40s -> 80s -> ....

Reply:
Ok. Let's do it that way. Do you think we should configure "max retries" or stick to "max delay"?

Reply:
Max retries make more sense to me.

Reply:
Ok, working on that. Thanks.